pkg/envfuncs/provider_funcs.go

```diff
@@ -0,0 +1,272 @@
+// for the generic handlers. Without this, making the pointer receiver is really painful or we have to
+// turn providers into value receivers and that doesn't really work well for the current implementation
+// we have for kind provider. xref: https://t.ly/GDxOB
+
+// Every time you add a new provider that implements the support.E2EClusterProvider interface, that
+// needs to be listed here in this interface type with properly or'ing that with the existing
+// values.
+//
+//	type provider interfae {
+//			kind.Cluster | k3d.Cluster | kwok.Cluster
+//	}
+type provider interface {
+	kind.Cluster
+}
+
+type providerRef[T any] interface {
+	*T
+	support.E2EClusterProvider
+}
+
+// GetClusterFromContext helps extract the E2EClusterProvider object from the context.
+// This can be used to setup and run tests of multi cluster e2e Prioviders.
+func GetClusterFromContext(ctx context.Context, clusterName string) (support.E2EClusterProvider, bool) {
+	c := ctx.Value(clusterNameContextKey(clusterName))
+	if c == nil {
+		return nil, false
+	}
+	cluster, ok := c.(support.E2EClusterProvider)
+	return cluster, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of GetClusterFromContext
+func GetKindClusterFromContext(ctx context.Context, clusterName string) (*kind.Cluster, bool) {
+	provider, ok := GetClusterFromContext(ctx, clusterName)
+	if ok {
+		return provider.(*kind.Cluster), ok
+	}
+	return nil, ok
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateCluster which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindCluster(clusterName string) env.Func {
+	return CreateCluster[kind.Cluster](clusterName)
+}
+
+// Deprecated: This handler has been deprecated in favor of CreateClusterWithConfig which can now accept
+// support.ClusterProvider type as input in order to setup the cluster using right providers
+func CreateKindClusterWithConfig(clusterName, image, configFilePath string) env.Func {
+	return CreateClusterWithConfig[kind.Cluster](clusterName, configFilePath, kind.WithImage(image))
+}
+
+// Deprecated: This handler has been deprecated in favor of DestroyCluster
+func DestroyKindCluster(name string) env.Func {
+	return DestroyCluster(name)
+}
+
+// Deprecated: This handler has been deprecated in favor of ExportClusterLogs
+func ExportKindClusterLogs(name, dest string) env.Func {
+	return ExportClusterLogs(name, dest)
+}
+
+// CreateCluster returns an env.Func that is used to
+// create an E2E provider cluster that is then injected in the context
+// using the name as a key.
+//
+// NOTE: the returned function will update its env config with the
+// kubeconfig file for the config client.
+func CreateCluster[T provider, P providerRef[T]](clusterName string) env.Func {
+	return func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+		var p P = new(T)
```

Comment:
If `T` is assignable to `P`, why not just use a single type to represent provider?
Maybe I am missing something.

Reply:
Ok, thanks for explaining. So that's the only way to implement pointer receiver with genericized types.

Reply:
@vladimirvivien  yes for now. We will revisit this again if there is a better way to achieve this for sure. I will ask around with some go experts that I know to see if they can help me to check if there is a better way.